lib/WUI/espif.c

```diff
@@ -0,0 +1,521 @@
+#include "string.h"
+#include "espif.h"
+#include "main.h"
+#include "stm32_port.h"
+#include "ff.h"
+#include "wui_api.h"
+
+#include "lwip/def.h"
+#include "lwip/ethip6.h"
+#include "lwip/etharp.h"
+
+#include "log.h"
+
+LOG_COMPONENT_DEF(ESPIF, LOG_SEVERITY_INFO);
+
+/*
+ * UART and other pin configuration for ESP01 module
+ *
+ * UART:                USART6
+ * STM32 TX (ESP RX):   GPIOC, GPIO_PIN_6
+ * STM32 RX (ESP TX):   GPIOC, GPIO_PIN_7
+ * RESET:               GPIOC, GPIO_PIN_13
+ * GPIO0:               GPIOE, GPIO_PIN_6
+ * GPIO2:               not connected
+ * CH_PD:               connected to board 3.3 V
+ *
+ * UART_DMA:           DMA2
+ * UART_RX_STREAM      STREAM_1
+ * UART_TX_STREAM      STREAM_6
+ */
+
+/*
+ * ESP UART NIC
+ *
+ * This provides a LwIP NIC implementation on top of a simple UART protocol used to communicate MAC address, link
+ * status and packets with ESP8266 attached on the other end of the UART. This requires custom FW running in the ESP
+ * implementing the protocol.
+ *
+ * Known issues:
+ * - This does not use netif state. All the state is kept in static varibles -> only on NIC is supported
+ *   (Maybe it is worh encapsulating the state just for the sake of code clarity.)
+ * - This runs at 1Mbaud even when ESP support 4.5Mbaud. There is some wierd coruption at higher speeds
+ *   (ESP seems to miss part of the packet data)
+ * - This does not effload checksum computation to ESP. Would be nice to enable parity and make the protocol more
+ *   robust (i.e using some counter to match packet begin and end while ensuring no data is lost). Provided UART
+ *   can be trusted not to alternate packet content the ESP would be able to compute packet checksums.
+ *
+ */
+
+// Thread stuffs
+static osThreadId UartBufferThread_id;
+static void uart_reader(void const *arg);
+osThreadDef(UartBufferThread, uart_reader, osPriorityBelowNormal, 0, 512);
+
+// Message queue stuffs
+osMessageQDef(uartBufferMbox, 16, NULL);
+static osMessageQId uartBufferMbox_id;
+
+// NIC state
+static uint32_t esp_working_mode;
+static uint32_t initialized;
+static bool associated = 0;
+static TaskHandle_t init_task_handle;
+
+// UART
+static const uint32_t ESPIF_UART_BAUDRATE = 1000000;
+uint8_t dma_buffer_rx[RX_BUFFER_LEN];
+static size_t old_dma_pos = 0;
+
+static void uart_input(uint8_t *data, size_t size, struct netif *netif);
+
+// UART Protocol
+static const char INTRON[4] = { 'U', 'N', 'U', 1 };
+
+enum MessageType {
+    MSG_DEVINFO = 0,
+    MSG_LINK = 1,
+    MSG_CLIENTCONFIG = 3,
+    MSG_PACKET = 4,
+};
+
+#define MAC_DATA_MAX_LEN 16
+
+void espif_set_operating_mode(uint32_t mode) {
+    esp_working_mode = mode;
+}
+
+uint32_t espif_get_operating_mode(void) {
+    return esp_working_mode;
+}
+
+void espif_receive_data(UART_HandleTypeDef *huart) {
+    LWIP_UNUSED_ARG(huart);
+    if (uartBufferMbox_id != NULL) {
+        uint32_t message = 0;
+        osMessagePut(uartBufferMbox_id, message, 0);
```

Comment:
Can some other thread delete the mbox between the check and now?

Reply:
This is no longer part of the source.